### ISSUE TITLE
[RFC] DDP should provide an option to ignore certain parameters

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#43690]
## 🚀 DDP should provide an option to ignore certain parameters

In certain use cases (described more below), we would like the option for users to be able to specify that certain parameters or modules within their network should be ignored by DDP. By ignored, we mean that DDP should not be aware of these parameters, i.e., not install/fire autograd hooks when these parameters receive gradient, not consider them when building gradient buckets, (all parameter tensors are used to control bucket sizing currently, for example), and not modify their grads. This will enable users to synchronize or otherwise manage these parameters that should be ignored in a way that is tailored to their use case.

### Current State
Currently, DDP recursively finds all parameters that require grad, and installs autograd hooks for all those parameters: [Code pointer 1](https://github.com/pytorch/pytorch/blob/79e6aaeb4c22ab6a9541ec547b9b24affe3896e7/torch/nn/parallel/distributed.py#L469) and [Code pointer 2](https://github.com/pytorch/pytorch/blob/79e6aaeb4c22ab6a9541ec547b9b24affe3896e7/torch/csrc/distributed/c10d/reducer.cpp#L97). This means that in general, we expect all parameters to be used in each backwards pass, and thus their autograd hook will fire and they will be scheduled for allreduce. The exception is if users pass in `find_unused_parameters=True`, errors will not be raised if certain parameters are not used in the forward pass of a certain iteration.

### Implementation proposal 
We can allow users to pass in a denylist set into DDP during construction time that allows them to specify modules/parameters that should be ignored by DDP and not have their autograd hooks installed. We can allow individual parameters, as well as `nn.Modules`. For `nn.Modules`, the semantics will be that all parameters that are part of the module will be ignored by DDP as well, recursively. 

Then, when finding all module parameters [here](https://github.com/pytorch/pytorch/blob/79e6aaeb4c22ab6a9541ec547b9b24affe3896e7/torch/nn/parallel/distributed.py#L465-L477), we can consult this list and omit parameters that are present in it. After this, the parameter should no longer be passed to the Reducer and thus things like bucketing and not installing autograd hooks should be accounted for.

One important point to mention is that this list of parameters should be the same across all ranks, otherwise, this could lead to desync during bucketing/allreduce. We should probably check this at construction time, it can likely be done as part of https://github.com/pytorch/pytorch/issues/32426 or an enhancement to it. 

Note that if the user ends up using these parameters anyways in the forward pass, we won't encounter errors but the user could perceive incorrectness since those params won't have their gradients synchronized. Users will have to be aware to keep what they pass in at construction time consistent with their model's forward pass.

### Code example
```
class TestModel(nn.Module):
            def __init__(self):
                super(TestModel, self).__init__()
                self.fc1 = nn.Linear(1, 1, bias=False)
                # Proxy that will be materialized to another architecture later.
                # (after wrapping model with DDP)
                self.fc2 = nn.Linear(10, 10, bias=False)

            def forward(self, x):
                x = self.fc1(x)
                x = self.fc2(x)
                return x

        device_id = self.rank
        model = TestModel().float().to(device_id)
        ddp = torch.nn.parallel.DistributedDataParallel(
            model,
            device_ids=[device_id],
           # Specify that we should ignore model.fc2 since it will be modified later
            params_and_modules_to_ignore=[model.fc2], # New argument of modules/params to ignore
        )
        # These params are not registered in DDP and thus don't
        # have autograd hooks etc installed on them.
        ddp.module.fc2 = nn.Linear(1, 1).to(device_id)
        inp = torch.rand([4, 1], dtype=torch.float)
        for i in range(6):
            ddp(inp).sum().backward()
            torch.cuda.synchronize(device=device_id)
```

## Motivation

This will help enable use cases such as hybrid parallelism where certain portions of the model do need dense parameter synchronization which is provided by DDP, but other portions of the model synchronize parameters differently. This could be achievable today by taking care to only wrap parts of the module that need DDP with it, although this could potentially split the overall model architecture into many DDP and non-DDP portions, which could hurt performance and is a poor developer experience.

## Alternatives

The alternative is to use `find_unused_parameters=True`. Under this, DDP will inspect outputs and find all tensors whose grad accumulators do not show up in the autograd graph (and thus will not get gradient). While this works, it is detrimental to performance since this requires the autograd graph search every iteration. If users know upfront that they don't want to use DDP allreduce on certain portions of the model globally, they should not have to pay this cost (which is more for use cases where the unused parameters can change from used to not used across iterations).

**Comment by qchip** [comment]
This reads great to me, and this feature will be very helpful for our hybrid distributed training paradigm, thank you very much! One issue relating to this but not exactly covered: our TorchScript model was generated from Python model from tracing, I am not sure why but somehow it has some unused parameters even after I removed those known unused parameters. It might not be trivial to find them, and clean them or add them to the skip list, in order to remove `find_unused_parameters=True`. It would be nice if we can automatically find and return the unused parameters, for example, during tracing, or report them when such assertion is raised with `find_unused_parameters=False`.

**Comment by xw285cornell** [comment]
This is awesome! Just to confirm: params_and_modules_to_ignore can take both the submodule and the parameters/tensors to ignore (so the typing will be something like List[Union[nn.Module, nn.Parameter]]), is that right?

**Comment by pritamdamania87** [comment]
Regarding `params_and_modules_to_ignore`, I'm wondering why we don't just have `params_to_ignore` and enforce that we have a list of parameters? It might be confusing for users how to use this API if they can pass in both parameters and modules. In addition to that, it's harder to maintain such an API since there are multiple ways this list can be built up and we have to consider several different options as part of the implementation.

**Comment by xw285cornell** [comment]
Agreed :)

**Comment by rohan-varma** [comment]
@pritamdamania87 @xw285cornell The argument can simply be `List[nn.Parameter]` as well, although  I proposed being able to accept nn.Module as well to make it a bit easier for the user, if we only accept parameters, the user will have to manually go through their modules and add all the parameters into a single list. With `module.parameters()` this is probably easy enough for the user, though.

**Comment by xw285cornell** [comment]
@rohan-varma any update on this? It seems we don't have much concern about this proposal :)

**Comment by rohan-varma** [comment]
@xw285cornell I'm beginning to now work on this and should have a PR out this week for it.

**Comment by xw285cornell** [comment]
Thanks @rohan-varma!
